### PR TITLE
2.next Support Stacking Flash Messages

### PR DESCRIPTION
**Pull request by xhs345**
In issue #7830 it was requested to backport the feature.

### CODE REVIEW

**Comment by markstory** [comment]
Looks good, I think changing the internal flash message format could be surprising in a bugfix release. This might have to wait until 2.10.0

**Comment by xhs345** [comment]
It's the nice kind of surprise, but I agree :)

**Comment by markstory** [comment]
I've retargeted this pull request to 2.next, but you may have to rebase to clean up the history.

**Comment by xhs345** [comment]
Branch has been rebased

File: lib/Cake/Controller/Component/FlashComponent.php [hunk 1]
```diff
@@ -82,12 +82,18 @@ public function set($message, $options = array()) {
 		}
 		$options['element'] = $plugin . 'Flash/' . $element;
 
-		CakeSession::write('Message.' . $options['key'], array(
+		$messages = CakeSession::read('Message.' . $options['key']);
+
+		$newMessage = array(
```
Comment (markstory) [inline lib/Cake/Controller/Component/FlashComponent.php]: In 3.x we support a `clear` option to allow for the flash message stack to be reset/cleared when adding a new message. Could we also have that option in 2.x? Without it maintaining existing behavior is not possible for those who want/need it.

**Review by markstory — CHANGES_REQUESTED** [review]

File: lib/Cake/Controller/Component/FlashComponent.php (hunk 1 above)
Comment (markstory) [inline lib/Cake/Controller/Component/FlashComponent.php]: We also have `clear` as a class level configuration option which makes it easy to preserve existing behavior through a configuration change.

**Comment by xhs345** [comment]
Yea, sounds like a good idea and was easy to add

File: lib/Cake/Controller/Component/FlashComponent.php [hunk 2]
```diff
@@ -38,6 +38,7 @@ class FlashComponent extends Component {
 		'key' => 'flash',
 		'element' => 'default',
 		'params' => array(),
+		'clear' => false
```
Comment (ravage84) [inline lib/Cake/Controller/Component/FlashComponent.php]: Haven't we agreed on a traling slash? :confused: 
Anyway, I would do so.

File: lib/Cake/View/Helper/FlashHelper.php [hunk 3]
```diff
@@ -82,10 +82,15 @@ public function render($key = 'flash', $options = array()) {
...
-		$flash = $options + $flash;
 		CakeSession::delete("Message.$key");
-		$flash['key'] = $key;
 
-		return $this->_View->element($flash['element'], $flash);
+		$out = '';
+		foreach ($flash as $message) {
+			$message['key'] = $key;
+			$message = $options + $message;
+			$out .= $this->_View->element($message['element'], $message);
```
Comment (ravage84) [inline lib/Cake/View/Helper/FlashHelper.php]: Nitpicking: Could we add a new line after/before each new element? Would produce nice HTML.

**Review by ravage84 — APPROVED** [review]
Looks good, just some nitpicking...

File: lib/Cake/View/Helper/FlashHelper.php (hunk 3 above)
Comment (xhs345) [inline lib/Cake/View/Helper/FlashHelper.php]: Right now the formatting is done in the view element and the HTML is looking quite nice by default imo 
screenshot 2016-11-01 08 57 40

File: lib/Cake/Controller/Component/FlashComponent.php (hunk 2 above)
Comment (ravage84) [inline lib/Cake/Controller/Component/FlashComponent.php]: Doh! Trailing comma, sorry :smirk_cat:

Comment (xhs345) [inline lib/Cake/Controller/Component/FlashComponent.php]: I don't know what you agreed on, but it looks like a trailing comma is not being used in similar instances the other components currently

File: lib/Cake/View/Helper/FlashHelper.php (hunk 3 above)
Comment (xhs345) [inline lib/Cake/View/Helper/FlashHelper.php]: Ok, guess I don't know which element and HTML you're referring to

**Pull request merged**